crates/chia-protocol/src/coin_record.rs

```diff
@@ -0,0 +1,49 @@
+use chia_streamable_macro::streamable;
+#[cfg(feature = "py-bindings")]
+use pyo3::pymethods;
+
+use crate::Coin;
+#[cfg(feature = "py-bindings")]
+use crate::{Bytes32, CoinState};
+
+#[streamable]
+#[derive(Copy)]
+pub struct CoinRecord {
+    // These are values that correspond to a CoinName that are used
+    // in keeping track of the unspent database.
+    coin: Coin,
+    confirmed_block_index: u32,
+    spent_block_index: u32,
+    coinbase: bool,
+    timestamp: u64, // Timestamp of the block at height confirmed_block_index
+}
+
+#[cfg(feature = "py-bindings")]
+#[pymethods]
+impl CoinRecord {
+    #[inline]
+    pub fn spent(&self) -> bool {
+        self.spent_block_index > 0
+    }
+
+    #[inline]
+    pub fn name(&self) -> Bytes32 {
+        self.coin.coin_id()
+    }
+
+    pub fn coin_state(&self) -> CoinState {
+        let spent_h = if self.spent_block_index > 0 {
+            Some(self.spent_block_index)
+        } else {
+            None
+        };
+
+        let confirmed_height = if self.confirmed_block_index == 0_u32 && self.timestamp == 0_u64 {
```

Comment:
I don't see what the timestamp has to do with the height - shouldn't this be the same as the check above?

Reply:
You can compare with `chia-blockchain` implementation this is 1:1. As the comment says `timestamp` is the timestamp in the block where the coin is confirmed, if it is not set then the coin is not confirmed

Reply:
the `timestamp` field on `CoinRecord` is a bit unfortunate I think. It's a short cut that mostly exists for this function (I believe). It would be possible to look up the timestamp of the block at `confirmed_block_index`, but we store this along with every coin instead.
In this context, "confirmed" means "created" right? Does `CoinRecord` support referring to coins that have not been created? I'm not so sure.

In fact, if you look in `chia-blockchain`, this function, `coin_state()` is *only* used in tests. I'm tempted to suggest we should remove it (and possibly put it in `chia/_tests/wallet/cat_wallet/test_cat_wallet.py`)

Reply:
I can't think of the scenario where the timestamp would be zero. Even reward coins are ascribed a timestamp of the next transaction block. It would be good to understand where this check comes from